src/main/java/org/elasticsearch/index/query/QueryParseContext.java

```diff
@@ -310,22 +313,45 @@ public Query parseInnerQuery() throws IOException, QueryParsingException {
         if (queryParser == null) {
             throw new QueryParsingException(index, "No query registered for [" + queryName + "]");
         }
-        Query result = queryParser.parse(this);
+        QueryBuilder result = queryParser.fromXContent(this);
         if (parser.currentToken() == XContentParser.Token.END_OBJECT || parser.currentToken() == XContentParser.Token.END_ARRAY) {
             // if we are at END_OBJECT, move to the next one...
             parser.nextToken();
         }
-        if (result instanceof NoCacheQuery) {
+        return result;
+    }
+
+    /**
+     * @deprecated replaced by calls to parseInnerQueryBuilder() and checkCachable() for the resulting queries
+     */
+    @Nullable
+    @Deprecated
+    public Query parseInnerQuery() throws IOException, QueryParsingException {
+        QueryBuilder builder = parseInnerQueryBuilder();
+
+        Query result = builder.toQuery(this);
+        checkCachable(result);
+        return result;
+    }
+
+    /**
+     * Checks if the given lucene query should be cached or wrapped, sets flags in this QueryParseContext accordingly
+     * @param query
+     */
+    public void checkCachable(Query query) {
```

Comment:
I will make it package private for now, will need to call this from toQuery() for nested Queries at the moment to stay consistend with existing code.